### PR TITLE
ci: release build with envoy standalone.

### PR DESCRIPTION
**Pull request by htuch**
This fixes an issue where the Envoy consumer git hash was picked up in the
release binary instead of Envoy proper.

### CODE REVIEW

**Comment by htuch** [comment]
@moderation @PiotrSikora @mattklein123

File: ci/do_ci.sh [hunk 1]
```diff
@@ -8,22 +8,21 @@ set -e
 echo "building using ${NUM_CPUS} CPUs"
 
 function bazel_release_binary_build() {
-  cd "${ENVOY_CONSUMER_SRCDIR}"
   echo "Building..."
-  bazel --batch build ${BAZEL_BUILD_OPTIONS} -c opt @envoy//source/exe:envoy-static.stripped.stamped
+  cd "${ENVOY_BUILD_DIR}"
```
Comment (PiotrSikora) [inline ci/do_ci.sh]: Do we need this? Shouldn't we build from within source (current) directory? Otherwise, this results in `Modified` git version, since none of the source files are in the tree.

Comment (htuch) [inline ci/do_ci.sh]: We need it for `bazel.coverage` builds, due to the weirdness of the circular symlinks that Bazel creates where you build and `gcovr` not being able to cope with this. For dev/release builds, we don't, so I will fix.

File: ci/do_ci.sh [hunk 2]
```diff
@@ -34,8 +33,8 @@ elif [[ "$1" == "bazel.asan" ]]; then
   cd "${ENVOY_CONSUMER_SRCDIR}"
   echo "Building and testing..."
   # TODO(htuch): This should switch to using clang when available.
-  bazel --batch test ${BAZEL_TEST_OPTIONS} -c dbg --config=asan --test_output=all \
-    --cache_test_results=no @envoy//test/... //:echo2_integration_test
+  bazel --batch test ${BAZEL_TEST_OPTIONS} -c dbg --config=asan @envoy//test/... \
```
Comment (PiotrSikora) [inline ci/do_ci.sh]: btw: we should use `@envoy//` prefix across the build, otherwise targets are going to be built twice - once for `//...` and once for `@envoy//...`, since Bazel thinks those are 2 different repositories.

See: https://github.com/bazelbuild/bazel/issues/1248

Comment (htuch) [inline ci/do_ci.sh]: I'm a bit confused, which of the labels do you want me to prefix with @envoy? @envoy is being used here to refer to the `local_repository`.

File: ci/do_ci.sh (hunk 1 above)
Comment (PiotrSikora) [inline ci/do_ci.sh]: OK, I don't get it.

It seems that removing `cd "${ENVOY_BUILD_DIR}"` should fix the issue, and indeed, calling `./bazel/get_workspace_status` from here yields `Clean` status, but the `@envoy//source/exe:envoy-static.stripped.stamped` binary still ends up having `Modified` build-id.

Running the same command from outside Docker results in `Clean` build-id.

Any ideas?

File: ci/do_ci.sh (hunk 2 above)
Comment (PiotrSikora) [inline ci/do_ci.sh]: Yes, but it also refers to the "current" workspace, since that's the name defined in `WORKSPACE`.

Although, now that I tested it, it seems that it doesn't work... I could swear that it did (some time ago), so either they regressed or I'm going crazy. See: https://github.com/bazelbuild/bazel/issues/1866

Anyway, please ignore this comment.

File: ci/do_ci.sh (hunk 1 above)
Comment (htuch) [inline ci/do_ci.sh]: Yeah, this is weird. It's even clean in the container immediately before/after the build, but `bazel-out/volatile-status.txt` says dirty. I'll take a look more in a bit, there's got to be a way.

https://github.com/bazelbuild/bazel/issues/2893 is tracking my concerns over the documentation and support-level for Bazel workspace status and C++. I think it's not release quality yet, much like the coverage - there's in theory a way to do it if you hack around enough and spend O(days) figuring it out yourself, but no documentation way of doing it or integration/regression tests.

Comment (PiotrSikora) [inline ci/do_ci.sh]: Should we merge this now or do you want to look into this today / tomorrow? Even with `Modified`, it's still better than what's currently in tree.

Comment (htuch) [inline ci/do_ci.sh]: Will look into it now, if we can't figure it out in O(1 hour) then we should merge.

File: ci/do_ci.sh [hunk 3]
```diff
@@ -8,48 +8,50 @@ set -e
...
 }
 
 if [[ "$1" == "bazel.release" ]]; then
   echo "bazel release build with tests..."
   bazel_release_binary_build
   echo "Testing..."
-  bazel --batch test ${BAZEL_TEST_OPTIONS} -c opt --test_output=all \
-    --cache_test_results=no @envoy//test/... //:echo2_integration_test
+  bazel --batch test ${BAZEL_TEST_OPTIONS} -c opt //test/...
+  rm -f bazel-*
```
Comment (PiotrSikora) [inline ci/do_ci.sh]: Why is this needed? It kills incremental builds.

Comment (htuch) [inline ci/do_ci.sh]: It was a small nod to the notion that someone might run `bazel.dev` and then `bazel.coverage`. The annoying circular symlinks will then cause problems for `bazel.coverage`. Let me move this into the `bazel.coverage` target to avoid it hurting incremental builds.

Comment (PiotrSikora) [inline ci/do_ci.sh]: Actually, incremental builds work fine, since it's just removing symlinks, not purging cache.

You can leave it as is.

Comment (PiotrSikora) [inline ci/do_ci.sh]: but yeah, it might be better served inside `bazel.coverage` target, in case someone runs `bazel.coverage` twice in a row.

File: ci/build_setup.sh [hunk 4]
```diff
@@ -83,3 +79,17 @@ cp -f "${ENVOY_SRCDIR}"/ci/WORKSPACE "${ENVOY_BUILD_DIR}"
...
+export ENVOY_CI_DIR="${ENVOY_SRCDIR}"/ci
+
+# Hack due to https://github.com/lyft/envoy/issues/838 and the need to have
+# tools and bazel.rc available for build linkstamping.
+mkdir -p "${ENVOY_CONSUMER_SRCDIR}"/tools
+mkdir -p "${ENVOY_CI_DIR}"/tools
+ln -sf "${ENVOY_SRCDIR}"/tools/bazel.rc "${ENVOY_CONSUMER_SRCDIR}"/tools/
+ln -sf "${ENVOY_SRCDIR}"/tools/bazel.rc "${ENVOY_CI_DIR}"/tools/
+mkdir -p "${ENVOY_CONSUMER_SRCDIR}"/bazel
+mkdir -p "${ENVOY_CI_DIR}"/bazel
```
Comment (PiotrSikora) [inline ci/build_setup.sh]: Actually, one more thing.

Could you remove `ci/{bazel,tools}` directories? or if you don't think that's valid option, then add them to `.gitignore`? Otherwise tree isn't clean.
